packages/flutter_tools/lib/src/fuchsia/application_package.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../application_package.dart';
+import '../base/file_system.dart';
+import '../build_info.dart';
+import '../project.dart';
+
+abstract class FuchsiaApp extends ApplicationPackage {
+  FuchsiaApp({@required String projectBundleId}) : super(id: projectBundleId);
+
+  /// Creates a new [FuchsiaApp] from a fuchsia sub project.
+  factory FuchsiaApp.fromFuchsiaProject(FuchsiaProject project) {
+    return BuildableFuchsiaApp(
+      project: project,
+    );
+  }
+
+  /// Creates a new [FuchsiaApp] from an existing farArchive.
+  ///
+  /// `applicationBinary` is the path to the farArchive.
+  factory FuchsiaApp.fromPrebuiltApp(FileSystemEntity applicationBinary) {
+    return PrebuiltFuchsiaApp(
+      farArchive: applicationBinary.path,
+    );
+  }
+
+  @override
+  String get displayName => id;
+
+  /// The location of the 'far' archive containing the built app.
+  String farArchive(BuildMode buildMode);
+}
+
+class PrebuiltFuchsiaApp extends FuchsiaApp {
+  PrebuiltFuchsiaApp({
+    @required String farArchive,
+  }) : _farArchive = farArchive,
+       super(projectBundleId: farArchive);
```

Comment:
Is the archive name really the right thing to use for the bundle id?  Seems like we should have an extra parameter here for the id.  Or maybe it doesn't matter?  Can we use null for the id?  Is the id even needed in the super class?

Reply:
At the moment, using the far archive name will likely confuse log parsing and isolate discovery, but this PR is already getting big, so I added a TODO to extract the app name from the archive as on other platforms.